### PR TITLE
arm: nxp: take PWM prescaler from device tree

### PR DESCRIPTION
**Pull request by bbilas**
Replace the hardcoded kPWM_Prescale_Divide_128
driver prescaler with the ones defined in the dtsi file that allow overriding them by the user.

Signed-off-by: Bartosz Bilas <b.bilas@grinn-global.com>

### CODE REVIEW

File: dts/arm/nxp/nxp_rt10xx.dtsi [hunk 1]
```diff
@@ -550,6 +550,7 @@
 				interrupts = <102 0>;
 				#pwm-cells = <2>;
 				clocks = <&ccm IMX_CCM_PWM_CLK 0 0>;
+				prescaler = <7>;
```
Comment (gmarull) [inline dts/arm/nxp/nxp_rt10xx.dtsi:553]: Similar to ST, we should use vendor prefix here, `nxp,prescaler`. @galak we should probably settle some rules on vendor-specific properties (we have many inconsistencies in-tree).

Comment (bbilas) [reply dts/arm/nxp/nxp_rt10xx.dtsi]: I have added the appropriate prefix.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 2]
```diff
@@ -256,6 +256,7 @@
 			interrupts = <83 0>, <84 0>;
 			#pwm-cells = <2>;
 			clocks = <&syscon MCUX_BUS_CLK 0 0>;
+			nxp,prescaler = <7>;
```
Comment (gmarull) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:259]: why is it defined at soc dts files? its value is really application specific (e.g. the pwm frequency you need to synthesize in your app), so it should be set in board files.

If register always has the same reset default, you can add such default to the property.

Comment (bbilas) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi:259]: I've followed how this is done for the rest vendors that have that property in the soc dts file.

File: dts/bindings/pwm/nxp,imx-pwm.yaml [hunk 3]
```diff
@@ -26,6 +26,22 @@ properties:
       description: |
         Enable for PWM to keep running in debug mode.
 
+    nxp,prescaler:
+      type: int
+      enum: [0, 1, 2, 3, 4, 5, 6, 7]
```
Comment (danieldegrasse) [inline dts/bindings/pwm/nxp,imx-pwm.yaml:31]: Why not use a string enum here, and use `DT_ENUM_IDX` to get the integer value to use for the prescaler property? Something like this: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/nxp%2Cmcux-rt-pinctrl.yaml#L79

Comment (bbilas) [reply dts/bindings/pwm/nxp,imx-pwm.yaml]: I didn't know that such macro. Fixed.

**Comment by bbilas** [comment]
CI failures seem to be not related to those changes.

**Review by danieldegrasse — APPROVED** [review]

File: drivers/pwm/pwm_mcux.c [hunk 4]
```diff
@@ -192,7 +192,7 @@ static const struct pwm_driver_api pwm_mcux_driver_api = {
 		.base = (PWM_Type *)DT_REG_ADDR(DT_INST_PARENT(n)),	  \
 		.index = DT_INST_PROP(n, index),			  \
 		.mode = kPWM_EdgeAligned,				  \
-		.prescale = kPWM_Prescale_Divide_128,			  \
+		.prescale = _CONCAT(kPWM_Prescale_Divide_, DT_INST_PROP(n, nxp_prescaler)),\
```
Comment (mmahadevan108) [inline drivers/pwm/pwm_mcux.c:195]: `DT_INST_ENUM_IDX` work better?

Comment (bbilas) [reply drivers/pwm/pwm_mcux.c:195]: Not really, this approach is bulletproof and we are sure that we set the correct value based on the `_pwm_clock_prescale` enum at the build time. There is a couple of implementation in the codebase that successfully uses this solution, see:

<pre>
zephyr/drivers/watchdog/wdt_mcux_wdog32.c:#define TO_WDOG32_CLK_SRC(val) _DO_CONCAT(kWDOG32_ClockSource, val)
zephyr/drivers/watchdog/wdt_mcux_wdog32.c:#define TO_WDOG32_CLK_DIV(val) _DO_CONCAT(kWDOG32_ClockPrescalerDivide, val)
zephyr/drivers/counter/counter_mcux_lptmr.c:#define TO_LPTMR_CLK_SEL(val) _DO_CONCAT(kLPTMR_PrescalerClock_, val)
zephyr/drivers/counter/counter_mcux_lptmr.c:#define TO_LPTMR_PIN_SEL(val) _DO_CONCAT(kLPTMR_PinSelectInput_, val)
zephyr/drivers/counter/counter_mcux_lptmr.c:#define TO_LPTMR_PRESCALER(val) _DO_CONCAT(LPTMR_PRESCALER_, val)
zephyr/drivers/counter/counter_mcux_lptmr.c:#define TO_LPTMR_GLITCH(val) _DO_CONCAT(LPTMR_GLITCH_, val)
zephyr/drivers/pwm/pwm_mcux_pwt.c:#define TO_PWT_PRESCALE_DIVIDE(val) _DO_CONCAT(kPWT_Prescale_Divide_, val)
zephyr/drivers/pwm/pwm_mcux_ftm.c:#define TO_FTM_PRESCALE_DIVIDE(val) _DO_CONCAT(kFTM_Prescale_Divide_, val)
zephyr/drivers/adc/adc_mcux_lpadc.c:	_DO_CONCAT(kLPADC_ReferenceVoltageAlt, val)
zephyr/drivers/adc/adc_mcux_lpadc.c:	_DO_CONCAT(kLPADC_ConversionAverage, val)
zephyr/drivers/adc/adc_mcux_lpadc.c:	_DO_CONCAT(kLPADC_PowerLevelAlt, val)
zephyr/drivers/adc/adc_mcux_adc12.c:#define TO_ADC12_CLOCK_SRC(val) _DO_CONCAT(kADC12_ClockSourceAlt, val)
zephyr/drivers/adc/adc_mcux_adc12.c:#define TO_ADC12_CLOCK_DIV(val) _DO_CONCAT(kADC12_ClockDivider, val)
zephyr/drivers/adc/adc_mcux_12b1msps_sar.c:#define TO_RT_ADC_CLOCK_DIV(val) _DO_CONCAT(kADC_ClockDriver, val)
zephyr/drivers/timer/mcux_lptmr_timer.c:#define TO_LPTMR_PRESCALER(val) _DO_CONCAT(LPTMR_PRESCALER_, val)
zephyr/drivers/timer/mcux_lptmr_timer.c:#define TO_LPTMR_CLK_SEL(val) _DO_CONCAT(kLPTMR_PrescalerClock_, val)
zephyr/drivers/dac/dac_mcux_dac.c:	_DO_CONCAT(kDAC_ReferenceVoltageSourceVref, val)
zephyr/drivers/dac/dac_mcux_dac32.c:	_DO_CONCAT(kDAC32_ReferenceVoltageSourceVref, val)
</pre>

**Review by mmahadevan108 — APPROVED** [review]

**Pull request merged**